pkg/apis/upgrade.cattle.io/v1/types.go

```diff
@@ -36,39 +40,70 @@ type Plan struct {
 	Secrets []SecretSpec `json:"secrets,omitempty"`
-
+	// Specify which node taints should be tolerated by pods applying the upgrade.
+	// Anything specified here is appended to the default of:
+	// - `{key: node.kubernetes.io/unschedulable, effect: NoSchedule, operator: Exists}`
 	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
-
+	// Jobs for exclusive plans cannot be run alongside any other exclusive plan.
 	Exclusive bool `json:"exclusive,omitempty"`
-
-	Window            *TimeWindowSpec               `json:"window,omitempty"`
-	Prepare           *ContainerSpec                `json:"prepare,omitempty"`
-	Cordon            bool                          `json:"cordon,omitempty"`
-	Drain             *DrainSpec                    `json:"drain,omitempty"`
-	Upgrade           *ContainerSpec                `json:"upgrade,omitempty" wrangler:"required"`
-	ImagePullSecrets  []corev1.LocalObjectReference `json:"imagePullSecrets,omitempty"`
-	PostCompleteDelay *metav1.Duration              `json:"postCompleteDelay,omitempty"`
+	// A time window in which to execute Jobs for this Plan.
+	// Jobs will not be generated outside this time window, but may continue executing into the window once started.
+	Window *TimeWindowSpec `json:"window,omitempty"`
+	// The prepare init container, if specified, is run before cordon/drain which is run before the upgrade container.
+	Prepare *ContainerSpec `json:"prepare,omitempty"`
+	// The upgrade container; must be specified.
+	Upgrade *ContainerSpec `json:"upgrade"`
+	// If Cordon is true, the node is cordoned before the upgrade container is run.
+	// If drain is specified, the value for cordon is ignored, and the node is cordoned.
+	// If neither drain nor cordon are specified and the node is marked as schedulable=false it will not be marked as schedulable=true when the Job completes.
+	Cordon bool `json:"cordon,omitempty"`
+	// Configuration for draining nodes prior to upgrade. If left unspecified, no drain will be performed.
+	Drain *DrainSpec `json:"drain,omitempty"`
+	// Image Pull Secrets, used to pull images for the Job.
+	ImagePullSecrets []corev1.LocalObjectReference `json:"imagePullSecrets,omitempty"`
+	// Time after a Job for one Node is complete before a new Job will be created for the next Node.
+	PostCompleteDelay *metav1.Duration `json:"postCompleteDelay,omitempty"`
 }
 
 // PlanStatus represents the resulting state from processing Plan events.
 type PlanStatus struct {
-	Conditions    []genericcondition.GenericCondition `json:"conditions,omitempty"`
-	LatestVersion string                              `json:"latestVersion,omitempty"`
-	LatestHash    string                              `json:"latestHash,omitempty"`
-	Applying      []string                            `json:"applying,omitempty"`
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []genericcondition.GenericCondition `json:"conditions,omitempty"`
```

Comment:
Doesn't need it, but not a bad idea to comment which conditions are expected and when.

Reply:
There are doc comments on the Condition vars up above but it appears those do not get used anywhere.